File: combinatorics/resource/ST/project_4/bai_toan_2/count_partitions_max.cpp
#include <iostream>
#include <vector>
using namespace std;

// Ham dem so phan hoach cua n voi so lon nhat la k
int countPartitionsMax(int n, int k) {
    // Tao bang 2 chieu dp[n+1][k+1], khoi tao toan bo = 0
    vector<vector<int>> dp(n + 1, vector<int>(k + 1, 0));

    // Co 1 cach de phan hoach 0 voi moi k
    for (int j = 0; j <= k; ++j)
        dp[0][j] = 1;

    // Duyet cac gia tri tu 1 den n va 1 den k
    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= k; ++j) {
            if (i - j >= 0)
                // Chon j: cong don khong chon j + chon j
                dp[i][j] = dp[i][j - 1] + dp[i - j][j];
            else
                // Khong the chon j vi lon hon i
                dp[i][j] = dp[i][j - 1];
        }
    }

    return dp[n][k];
}

int main() {
    int n, k;
    cout << "Nhap n: ";
    cin >> n;
    cout << "Nhap k: ";
    cin >> k;

    // In ket qua
    cout << "So phan hoach cua " << n << " voi so lon nhat la " << k
         << " la: " << countPartitionsMax(n, k) << endl;

    return 0;
}

File: combinatorics/resource/ST/project_4/bai_toan_2/count_partitions_max.py
def count_partitions_max(n, k):
    # Tao bang dp voi n+1 hang va k+1 cot, khoi tao bang 0
    dp = [[0] * (k + 1) for _ in range(n + 1)]

    # Co 1 cach phan hoach 0 voi moi k (khong chia gi ca)
    for j in range(k + 1):
        dp[0][j] = 1

    # Duyet qua cac gia tri i tu 1 den n
    for i in range(1, n + 1):
        for j in range(1, k + 1):
            if i - j >= 0:
                # Co 2 truong hop: khong chon j va chon j
                dp[i][j] = dp[i][j - 1] + dp[i - j][j]
            else:
                # Chi co the khong chon j
                dp[i][j] = dp[i][j - 1]

    return dp[n][k]

# Nhap n va k tu ban phim
n = int(input("Nhap n: "))
k = int(input("Nhap k: "))

# In ket qua ra man hinh
print(f"So phan hoach cua {n} voi so lon nhat la {k} la: {count_partitions_max(n, k)}")